models/migrations/v1_22/v283.go

```diff
@@ -0,0 +1,350 @@
+					docsEntries[1] = entry
+				}
+			case ".github":
+				if entry.Name() == ".github" || docsEntries[2] == nil {
+					docsEntries[2] = entry
+				}
+			}
+			continue
+		}
+		if i, ok := util.IsFileExtension(entry.Name(), fileType, exts...); ok {
+			log.Debug("Potential %s file: %s", fileType, entry.Name())
+			if targetFiles[i] == nil || base.NaturalSortLess(targetFiles[i].Name(), entry.Blob().Name()) {
+				if entry.IsLink() {
+					target, err := entry.FollowLinks()
+					if err != nil && !git.IsErrBadLink(err) {
+						return "", nil, err
+					} else if target != nil && (target.IsExecutable() || target.IsRegular()) {
+						targetFiles[i] = entry
+					}
+				} else {
+					targetFiles[i] = entry
+				}
+			}
+		}
+	}
+	var targetFile *git.TreeEntry
+	for _, f := range targetFiles {
+		if f != nil {
+			targetFile = f
+			break
+		}
+	}
+
+	if treePath == "" && targetFile == nil {
+		for _, subTreeEntry := range docsEntries {
+			if subTreeEntry == nil {
+				continue
+			}
+			subTree := subTreeEntry.Tree()
+			if subTree == nil {
+				// this should be impossible; if subTreeEntry exists so should this.
+				continue
+			}
+			var err error
+			childEntries, err := subTree.ListEntries()
+			if err != nil {
+				return "", nil, err
+			}
+
+			subfolder, targetFile, err := findFileInEntries(fileType, childEntries, subTreeEntry.Name(), language, false)
+			if err != nil && !git.IsErrNotExist(err) {
+				return "", nil, err
+			}
+			if targetFile != nil {
+				return path.Join(subTreeEntry.Name(), subfolder), targetFile, nil
+			}
+		}
+	}
+
+	return "", targetFile, nil
+}
+
+func localizedExtensions(ext, languageCode string) (localizedExts []string) {
+	if len(languageCode) < 1 {
+		return []string{ext}
+	}
+
+	lowerLangCode := "." + strings.ToLower(languageCode)
+
+	if strings.Contains(lowerLangCode, "-") {
+		underscoreLangCode := strings.ReplaceAll(lowerLangCode, "-", "_")
+		indexOfDash := strings.Index(lowerLangCode, "-")
+		// e.g. [.zh-cn.md, .zh_cn.md, .zh.md, _zh.md, .md]
+		return []string{lowerLangCode + ext, underscoreLangCode + ext, lowerLangCode[:indexOfDash] + ext, "_" + lowerLangCode[1:indexOfDash] + ext, ext}
+	}
+
+	// e.g. [.en.md, .md]
+	return []string{lowerLangCode + ext, ext}
+}
+
+// detectLicenseByEntry returns the licenses detected by the given tree entry
+func detectLicenseByEntry(file *git.TreeEntry) ([]string, error) {
+	if file == nil {
+		return nil, nil
+	}
+
+	blob := file.Blob()
+	content, err := blob.GetBlobContent(setting.UI.MaxDisplayFileSize)
+	if err != nil {
+		return nil, fmt.Errorf("GetBlobAll: %w", err)
+	}
+	return detectLicense(content), nil
+}
+
+// detectLicense returns the licenses detected by the given content buff
+func detectLicense(content string) []string {
+	if len(content) == 0 {
+		return nil
+	}
+	if classifier == nil {
+		log.Error("detectLicense: license classifier is null.")
+		return nil
+	}
+
+	matches, err := classifier.MatchFrom(strings.NewReader(content))
```

Comment:
Should just get the content reader from the blob and pass it to the `classifier` without copy into memory.

Reply:
Done in https://github.com/go-gitea/gitea/pull/24872/commits/dac32cede886c5a070e06f8ccf39748918b649e5.